File: src/core/Shared/TradeGrade.cs
using System;

namespace core.Shared
{
    public readonly struct TradeGrade : IComparable
    {
        public TradeGrade(string grade)
        {
            if (string.IsNullOrWhiteSpace(grade))
            {
                throw new ArgumentException("Grade cannot be blank", nameof(grade));
            }
            
            // right now we allow only A, B, C
            // this feels like should be configurable by trader?
            if (!grade.Equals("A", StringComparison.OrdinalIgnoreCase) &&
                !grade.Equals("B", StringComparison.OrdinalIgnoreCase) &&
                !grade.Equals("C", StringComparison.OrdinalIgnoreCase))
            {
                throw new ArgumentException("Grade must be A, B, or C", nameof(grade));
            }

            Value = grade.ToUpper();
        }

        public string Value { get; }

        public int CompareTo(object obj)
        {
            if (obj is TradeGrade other)
            {
                return string.Compare(Value, other.Value, StringComparison.Ordinal);
            }

            throw new ArgumentException("Object is not a TradeGrade");
        }
    }
}

File: src/core2.fs/Shared/TradeGrade.fs
namespace core.Shared

open System

[<CustomComparison; StructuralEquality>]
[<Struct>]
type TradeGrade =
    val Value: string
    
    new(grade: string) =
        if String.IsNullOrWhiteSpace(grade) then
            invalidArg "grade" "Grade cannot be blank"
        
        // right now we allow only A, B, C
        // this feels like should be configurable by trader?
        let upper = grade.ToUpper()
        if upper <> "A" && upper <> "B" && upper <> "C" then
            invalidArg "grade" "Grade must be A, B, or C"
        
        { Value = upper }
    
    interface IComparable with
        member this.CompareTo(obj: obj) =
            match obj with
            | :? TradeGrade as other -> String.Compare(this.Value, other.Value, StringComparison.Ordinal)
            | _ -> invalidArg "obj" "Object is not a TradeGrade"
